cgi-bin/LJ/Talk.pm

```diff
@@ -2831,6 +2831,270 @@ sub prepare_and_validate_comment {
+        return $mlerr->("$SC.error.confused_identity");
+    };
+
+    # just so we only have to check this once.
+    unless (ref $didlogin eq 'SCALAR') {
+        my $throwaway = 0;
+        $didlogin = \$throwaway;
+    }
+
+    # User stuff!
+    # usertype - One of the following, with the following extra fields
+        # anonymous
+            # nothing
+        # openid
+            # oidurl
+            # oiddo_login
+        # openid_cookie
+            # nothing, OR:
+            # cookieuser (= ext_1234) (QR)
+        # cookieuser (currently logged in user)
+            # cookieuser (= username)
+        # user (non-logged-in user, w/ name/password provided in the form)
+            # userpost - The username provided in the form
+            # password
+
+    # CHECKLIST:
+    # 1. Check for incoherent combinations of fields. Most of these can only
+    # happen if javascript is disabled. From what I've been told, there used to
+    # be rare conditions where it would straight-up post as the wrong user; who
+    # knows if that's somehow still true, but regardless, if we got conflicting
+    # information, then the user's intention was not clear and we need to ask
+    # them to clarify, because they might have meant it either way.
+    # 2. Validate the specified user type's credentials.
+    # 3. If validated, return the LJ::User object for that user.
+    if ( $form->{usertype} eq 'anonymous' ) {
+        if ($form->{oidurl} || $form->{userpost}) {
+            return $incoherent->();
+        }
+        return undef; # Well! that was easy.
+    } elsif ( $form->{usertype} eq 'cookieuser' ) {
+        if ($form->{oidurl}) {
+            return $incoherent->();
+        }
+        # If they selected "current user" and then typed in their own username,
+        # well, that's "wrong" but their intention was perfectly clear. But if
+        # they typed in a DIFFERENT username, get outta here.
+        if ( $form->{userpost} && ($form->{userpost} ne $form->{cookieuser}) ) {
+            return $incoherent->();
+        }
+
+        # OK! Check if that's the logged-in user.
+        if ( $remote && ($remote->user eq $form->{cookieuser}) ) {
+            return $remote; # Cool.
+        } else {
+            return $mlerr->("$SC.error.lostcookie");
+        }
+    } elsif ( $form->{usertype} eq 'user' ) {
+        if ($form->{oidurl}) {
+            return $incoherent->();
+        }
+        # No username?
+        if ( ! $form->{userpost} ) {
+            my $iscomm = $journalu->is_community ? '.comm' : '';
+            my $noanon = $journalu->prop('opt_whocanreply') eq 'all' ? '' : '.noanon';
+            return $mlerr->( "$SC.error.nousername$noanon$iscomm", { sitename => $LJ::SITENAMESHORT } );
+        }
+
+        my $exptype;    # set to long if ! after username
+        my $ipfixed;    # set to remote  ip if < after username
+
+        # Parse inline login options.
+        # MUTATE FORM: remove trailing garbage from username.
```

Comment:
I really wonder if anybody actually uses this and would notice if it...magically stopped working. Besides me, anyway. Well, no reason to poke that horse I suppose.

Reply:
Put a counter on it, see how many ticks it tickies?